### PR TITLE
Adding event-handling doc

### LINKED ISSUE #1028: Add eventhandler example in user-guide
Can someone give me a quick example of what is meant here?
https://github.com/operator-framework/operator-sdk/blob/c5cc09b184ac4782c68b071a9da20d9ea2a9052d/doc/user-guide.md#L150
https://github.com/operator-framework/operator-sdk/blob/c5cc09b184ac4782c68b071a9da20d9ea2a9052d/doc/user-guide.md#L152

I think event handling is a very important topic that should be understood and directly shown to every new operator developer when reading the user guide.

Thanks! :D

### PR DESCRIPTION
**Pull request by gaurav-dalvi**
Added explanation and exmaples on how to use
1:EnqueueRequestForObject
2:EnqueueRequestForOwner
3:EnqueueRequestsFromMapFunc

Fixes #1028

### CODE REVIEW

**Comment by gaurav-dalvi** [comment]
@joelanford : Finally added the doc.

**Comment by gaurav-dalvi** [comment]
@LiliC

**Comment by hasbro17** [comment]
@gaurav-dalvi Thanks for the PR.

I've been taking a look again at upstream controller-runtime docs for the [event handlers pkg](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/handler).

It seems to me that the documentation for each type of handler is already clear and concise enough in the godocs. Having that here in the SDK repo again would at best be repeating/rephrasing the upstream docs. I think I mentioned in a [previous comment](https://github.com/operator-framework/operator-sdk/issues/1028#issuecomment-460439705) that we probably shouldn't repeat the upstream docs, but rather point out that there are different options for handler types and then just link to the upstream godocs.

However there certainly is room for improvement in illustrating the [example code for the EnqueueRequestsFromMapFunc](https://godoc.org/sigs.k8s.io/controller-runtime/pkg/handler#example-EnqueueRequestsFromMapFunc) handler.

And I think that was my main motivation for this TODO. With a more detailed example we can highlight how the `EnqueueRequestsFromMapFunc` handler can be useful for operators that have dependencies on resources that they don't own.

That can be done through your example. Or another example that I've seen in the wild for the [gardner-extensions handler](https://github.com/gardener/gardener-extensions/blob/d92672715480a1ff21fa64708be0908545b3ce0d/pkg/controller/controlplane/mapper.go#L37-L68) where a watch event for a secret is fanned out all the CRs that are using it in the spec.

**TLDR:** I would suggest that we update this PR to only point to the upstream godocs for now. And then try to update the examples and godocs on the controller-runtime where lacking.

**Comment by gaurav-dalvi** [comment]
Thanks @hasbro17 . @joelanford  and me talked on operator-sdk slack channel and then came up with this. My experience in working with operator-sdk for 2-3 months tells me that Memcache example is too naive and will never get used in most of the practical implementation of operator-SDKs. It took a significant amount of time to understand the usage of `EnqueueRequestsFromMapFunc ` . GoDo of that is limited and as you pointed out example is also not useful. Purpose of this PR is to make use-guide single page to understand the usage of all the handlers, which is the most used thing in operator-sdk by developers.  With practical example as per this PR makes much more easy to understand the use case of `EnqueueRequestsFromMapFunc` . Thats why i feel this PR is more relevant in the context of operator-sdk.

**Comment by hasbro17** [comment]
The memcached example is suitable for the purposes of the user-guide, which is to demonstrate the basic workflow and APIs.
The intent is not to have users use that as a base for building their projects.

The issue I have with the doc in this PR is that the explanations and examples for `EnqueueRequestForObject` and `EnqueueRequestForOwner` are much less precise than the ones in the usptream godocs. That's why I would prefer to just link directly to them.

For example, just compare what we have here vs what's upstream for `EnqueueRequestForOwner`.
- Our doc:
  ```
  This is generally used in handlling secondary resource of the operator. It will map each event to 
  `Reconcile()` request for the owner of deployment.
  ```
- upstream godocs:
  ```
  EnqueueRequestForOwner enqueues Requests for the Owners of an object. E.g. the object that 
  created the object that was the source of the Event.
  
  If a ReplicaSet creates Pods, users may reconcile the ReplicaSet in response to Pod Events 
  using:
  
  - a source.Kind Source with Type of Pod.

  - a handler.EnqueueRequestForOwner EventHandler with an OwnerType of ReplicaSet and 
  IsController set to true.
  ```

So at the very least I would not want to repeat the docs/examples for `EnqueueRequestForObject` and `EnqueueRequestForOwner` here in the SDK docs.

> Purpose of this PR is to make use-guide single page to understand the usage of all the handlers

If we're already referring to upstream docs for two of the handlers then that's not the case anymore. So that only leaves the `EnqueueRequestsFromMapFunc`.

> With practical example as per this PR makes much more easy to understand the use case of EnqueueRequestsFromMapFunc

Any example like the one that you've provided is going to be limited in isolation. In the case of your example it's just filtering pods events by checking if they are in the CR namespace.
It may be more practical to you since it might be from a familiar use case but it wasn't obvious to me why an operator would do that. For me it made much more sense for an operator to perhaps filter and fan out watch events for secrets that it's CRs depends on. So no example is going to be perfect.

Regardless, my point still stands that we can take this example and improve the upstream godoc since we expect users to look at that anyway.

And to further convey how it's used practically, we can point to actual operators that are using this handler so users can see the example in its full context.

In the end I would much rather improve the API docs upstream so it's more clear not just for SDK users, but for controller-runtime users in general.

**Comment by hasbro17** [comment]
@gaurav-dalvi I've added the links in the user-guide to the upstream EventHandler godocs in https://github.com/operator-framework/operator-sdk/pull/1783.
So I think we can close this for now, and try to submit a PR upstream to elaborate on the `EnqueueRequestsFromMapFunc` example.

**Comment by hasbro17** [comment]
Closing in favor of #1783

**Comment by gaurav-dalvi** [comment]
Thanks @hasbro17 . I will create PR against this repo to explain `EnqueueRequestsFromMapFunc ` with an example.
